rp2040-hal/src/clocks/mod.rs

```diff
@@ -22,21 +51,108 @@ impl ShareableClocks {
     }
 }
 
+const XOSC_MHZ: u32 = 12_000_000_u32;
+
+fn make_div<S: TryInto<Hertz<u32>>, F: TryInto<Hertz<u32>>>(
+    src_freq: S,
+    freq: F,
+) -> Result<u32, ()> {
+    let src_freq = u64::from(*src_freq.try_into().map_err(|_| ())?.integer());
+    let freq = (*freq.try_into().map_err(|_| ())?.integer()).into();
+    let div: u64 = (src_freq << 8).wrapping_div(freq);
+    Ok(div as u32)
+}
 /// Abstraction layer providing Clock Management.
 pub struct ClocksManager {
     clocks: CLOCKS,
     shared_clocks: ShareableClocks,
 }
 impl ClocksManager {
     /// Exchanges CLOCKS block against Self.
-    pub fn new(mut clocks_block: CLOCKS) -> Self {
+    pub fn new(mut clocks_block: CLOCKS, watchdog: &mut Watchdog) -> Self {
+        // Start tick in watchdog
+        watchdog.enable_tick_generation(XOSC_MHZ as u8);
+
+        // Disable resus that may be enabled from previous software
+        clocks_block.clk_sys_resus_ctrl.write_with_zero(|w| w);
+
         let shared_clocks = ShareableClocks::new(&mut clocks_block);
         ClocksManager {
             clocks: clocks_block,
             shared_clocks,
         }
     }
 
+    /// Initialize the clocks
+    pub fn init(
+        &self,
+        _: &CrystalOscillator<Stable>,
+        _: &PhaseLockedLoop<Locked, PLL_SYS>,
+        _: &PhaseLockedLoop<Locked, PLL_USB>,
+    ) {
+        // Configure clocks
+        // CLK_REF = XOSC (12MHz) / 1 = 12MHz
+        let mut ref_clock = self.ref_clock();
+        let div = make_div(12u32.MHz(), 12u32.MHz()).unwrap();
+        if div > ref_clock.get_div() {
+            ref_clock.set_div(div);
+        }
+        ref_clock.set_xosc_src();
+        ref_clock.await_select(2);
+        ref_clock.set_div(div);
+
+        // CLK SYS = PLL SYS (125MHz) / 1 = 125MHz
+        let mut sys_clock = self.sys_clock();
+        let div = make_div(125u32.MHz(), 125u32.MHz()).unwrap();
+        if div > sys_clock.get_div() {
+            sys_clock.set_div(div);
+        }
+        sys_clock.set_pll_sys_auxsrc();
+        sys_clock.set_self_aux_src();
+        sys_clock.await_select(1);
+        sys_clock.set_div(div);
+
+        // CLK USB = PLL USB (48MHz) / 1 = 48MHz
+        let mut usb_clock = self.usb_clock();
+        let div = make_div(48u32.MHz(), 48u32.MHz()).unwrap();
+        if div > usb_clock.get_div() {
+            usb_clock.set_div(div);
+        }
+        usb_clock.disable();
+        usb_clock.set_pll_usb_auxsrc();
+        usb_clock.enable();
+        usb_clock.set_div(div);
+
+        // CLK ADC = PLL USB (48MHZ) / 1 = 48MHz
+        let mut adc_clock = self.adc_clock();
+        let div = make_div(48u32.MHz(), 48u32.MHz()).unwrap();
+        if div > adc_clock.get_div() {
+            adc_clock.set_div(div);
+        }
+        adc_clock.disable();
+        adc_clock.set_pll_usb_auxsrc();
+        adc_clock.enable();
```

Comment:
Ah, wouldn't you want to enable the ADC clock inside the ADC peripheral driver?
Same with the other endpoint clocks here. Most HALs will enable the clock on peripheral creation, where you have the most knowledge about how the system is going to be used, and where an error would lead you to the correct place.

Reply:
Yes definitely, however I started with this to scratch my own itch and to get something out of the temp sensor. The adc is quite easy to do with just the pac. But for the clocks I basically used the C sdk defaults.  

I feel the clock subsystem needs a lot of rework to support this. But by adding the missing clocks and providing a somewhat sane default at least the rest can work..